src/theory/strings/theory_strings.h

```diff
@@ -612,7 +651,230 @@ class TheoryStrings : public Theory {
     ~Statistics();
   };/* class TheoryStrings::Statistics */
   Statistics d_statistics;
-  
+
+ private:
+  //-----------------------inference steps
+  /** check initial
+   *
+   * This function initializes term indices for each strings function symbol.
+   * One key aspect of this construction is that concat terms are indexed by
+   * their list of non-empty components. For example, if x = "" is an equality
+   * asserted in this SAT context, then y ++ x ++ z may be indexed by (y,z).
+   * This method may infer various facts while building these term indices, for
+   * instance, based on congruence. An example would be inferring:
+   *   y ++ x ++ z = y ++ z
+   * in this example, if both terms are registered in this SAT context.
+   *
+   * This function should be called as a first step of any strategy.
```

Comment:
We could add an Assert in addStrategyStep for this.

Reply:
Change looks good. Note that the assertion only checks that if `CHECK_INIT` is used, it is used in the first step. It does not check that the first step is a `CHECK_INIT`. The comment seems to suggest that there should be an additional assert `!d_infer_steps.empty() || s == CHECK_INIT`.

Reply:
I think it can be if-and-only-if, I've updated.